File: src/src/main/resources/mmlspark/_FindBestModel.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class _FindBestModel(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEstimator):
    """


    Args:

        evaluationMetric (str): Metric to evaluate models with (default: accuracy)
        models (object): List of models to be evaluated
    """

    @keyword_only
    def __init__(self, evaluationMetric="accuracy", models=None):
        super(_FindBestModel, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.FindBestModel")
        self._cache = {}
        self.evaluationMetric = Param(self, "evaluationMetric", "evaluationMetric: Metric to evaluate models with (default: accuracy)")
        self._setDefault(evaluationMetric="accuracy")
        self.models = Param(self, "models", "models: List of models to be evaluated", generateTypeConverter("models", self._cache, complexTypeConverter))
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, evaluationMetric="accuracy", models=None):
        """
        Set the (keyword only) parameters

        Args:

            evaluationMetric (str): Metric to evaluate models with (default: accuracy)
            models (object): List of models to be evaluated
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setEvaluationMetric(self, value):
        """

        Args:

            evaluationMetric (str): Metric to evaluate models with (default: accuracy)

        """
        self._set(evaluationMetric=value)
        return self


    def getEvaluationMetric(self):
        """

        Returns:

            str: Metric to evaluate models with (default: accuracy)
        """
        return self.getOrDefault(self.evaluationMetric)


    def setModels(self, value):
        """

        Args:

            models (object): List of models to be evaluated

        """
        self._set(models=value)
        return self


    def getModels(self):
        """

        Returns:

            object: List of models to be evaluated
        """
        return self._cache.get("models", None)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.FindBestModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=_FindBestModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".FindBestModel"
        return from_java(java_stage, module_name)

    def _create_model(self, java_model):
        return _BestModel(java_model)


class _BestModel(ComplexParamsMixin, JavaModel, JavaMLWritable, JavaMLReadable):
    """
    Model fitted by :class:`_FindBestModel`.

    This class is left empty on purpose.
    All necessary methods are exposed through inheritance.
    """

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.BestModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=_BestModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".BestModel"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/FindBestModel.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Microsoft.Spark.Interop.Internal.Java.Util;
using Microsoft.Spark.Utils;

using SynapseML.Dotnet.Utils;
using Synapse.ML.Automl;

namespace Synapse.ML.Automl
{
    /// <summary>
    /// <see cref="FindBestModel"/> implements FindBestModel
    /// </summary>
    public class FindBestModel : JavaEstimator<BestModel>, IJavaMLWritable, IJavaMLReadable<FindBestModel>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.automl.FindBestModel";

        /// <summary>
        /// Creates a <see cref="FindBestModel"/> without any parameters.
        /// </summary>
        public FindBestModel() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="FindBestModel"/> with a UID that is used to give the
        /// <see cref="FindBestModel"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public FindBestModel(string uid) : base(s_className, uid)
        {
        }

        internal FindBestModel(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets evaluationMetric value for <see cref="evaluationMetric"/>
        /// </summary>
        /// <param name="evaluationMetric">
        /// Metric to evaluate models with
        /// </param>
        /// <returns> New FindBestModel object </returns>
        public FindBestModel SetEvaluationMetric(string value) =>
            WrapAsFindBestModel(Reference.Invoke("setEvaluationMetric", (object)value));
        
        /// <summary>
        /// Sets models value for <see cref="models"/>
        /// </summary>
        /// <param name="models">
        /// List of models to be evaluated
        /// </param>
        /// <returns> New FindBestModel object </returns>
        public FindBestModel SetModels(JavaTransformer[] value)
        {
            var arrayList = new ArrayList(SparkEnvironment.JvmBridge);
            foreach (var v in value)
            {
                arrayList.Add(v);
            }
            return WrapAsFindBestModel(Reference.Invoke("setModels", (object)arrayList));
        }

        
        /// <summary>
        /// Gets evaluationMetric value for <see cref="evaluationMetric"/>
        /// </summary>
        /// <returns>
        /// evaluationMetric: Metric to evaluate models with
        /// </returns>
        public string GetEvaluationMetric() =>
            (string)Reference.Invoke("getEvaluationMetric");
        
        
        /// <summary>
        /// Gets models value for <see cref="models"/>
        /// </summary>
        /// <returns>
        /// models: List of models to be evaluated
        /// </returns>
        public JavaTransformer[] GetModels()
        {
            JvmObjectReference[] jvmObjects = (JvmObjectReference[])Reference.Invoke("getModels");
            JavaTransformer[] result = new JavaTransformer[jvmObjects.Length];
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            for (int i=0; i < jvmObjects.Length; i++)
            {
                if (JvmObjectUtils.TryConstructInstanceFromJvmObject(
                    jvmObjects[i],
                    classMapping,
                    out JavaTransformer instance))
                {
                    result[i] = instance;
                }
            }
            return result;
        }

        /// <summary>Fits a model to the input data.</summary>
        /// <param name="dataset">The <see cref="DataFrame"/> to fit the model to.</param>
        /// <returns><see cref="BestModel"/></returns>
        override public BestModel Fit(DataFrame dataset) =>
            new BestModel(
                (JvmObjectReference)Reference.Invoke("fit", dataset));

        /// <summary>
        /// Loads the <see cref="FindBestModel"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="FindBestModel"/> was saved to</param>
        /// <returns>New <see cref="FindBestModel"/> object, loaded from path.</returns>
        public static FindBestModel Load(string path) => WrapAsFindBestModel(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<FindBestModel> Read() =>
            new JavaMLReader<FindBestModel>((JvmObjectReference)Reference.Invoke("read"));

        private static FindBestModel WrapAsFindBestModel(object obj) =>
            new FindBestModel((JvmObjectReference)obj);

        
    }
}

        